src/primitives.cr

```diff
@@ -49,6 +49,56 @@ class Reference
   @[Primitive(:object_id)]
   def object_id : UInt64
   end
+
+  # Performs basic initialization so that the given *address* is ready for use
+  # as an object's instance data. Returns *address* cast to `self`'s type.
+  #
+  # More specifically, this is the part of object initialization that occurs
+  # after memory allocation and before calling one of the `#initialize`
+  # overloads. It zeroes the memory, sets up the type ID (necessary for dynamic
+  # dispatch), and then runs all inline instance variable initializers.
+  #
+  # *address* must point to a suitably aligned buffer of at least
+  # `instance_sizeof(self)` bytes.
+  #
+  # WARNING: This method is unsafe, as it assumes the caller is responsible for
+  # managing the memory at the given *address* manually.
+  #
+  # ```
+  # class Foo
+  #   getter i : Int64
+  #   getter str = "abc"
+  #
+  #   def initialize(@i)
+  #   end
+  #
+  #   foo_buffer = LibC.malloc(24)
+  #   foo = pre_initialize(foo)
+  #   foo.i                  # => 0
+  #   foo.str                # => "abc"
+  #   (foo || "").is_a?(Foo) # => true
+  #
+  #   foo.initialize(123_i64)
+  #   foo.i # => 123
+  # end
+  # ```
+  #
+  # See also: `Reference.unsafe_construct`.
+  @[Experimental("This API is still under development. Join the discussion about custom reference allocation at [#13481](https://github.com/crystal-lang/crystal/issues/13481).")]
+  @[Primitive(:pre_initialize)]
+  {% if compare_versions(Crystal::VERSION, "1.2.0") >= 0 %}
+    def self.pre_initialize(address : Pointer)
+      # This ensures `.pre_initialize` is instantiated for every subclass,
+      # otherwise all calls will refer to the sole instantiation in
+      # `Reference.class`. This is necessary when the receiver is a virtual
+      # metaclass type. Apparently this works even for primitives
+      {% type %}
```

Comment:
Not entirely sure what to do with this. A fresh compiler will break if the existing compiler's version is below 1.2.0 and the compiler itself uses `.pre_initialize` / `.unsafe_construct` on a virtual metaclass. I don't think that is very likely, but if we must do it, the only solution seems to be explicitly defining `.pre_initialize` in _every_ subclass of `Reference` in the compiler, similar to `.allocate`.